### ISSUE TITLE
ENH/BUG: Pandas nullable dtypes should be opt in

### ISSUE THREAD

**Issue by lithomas1** [dask/fastparquet#642]
fastparquet currently automatically uses the nullable dtypes when reading parquet files. In pandas, however, because the nullable dtypes are experimental in pandas, ``pandas.read_parquet`` uses a keyword called ``use_nullable_dtypes`` to allow users to opt-in/out of using the nullable dtypes. So, because the nullable dtypes are automatically used, some pandas tests are broken by fastparquet 0.7.0.

To fix this, fastparquet should also expose a ``use_nullable_dtypes`` keyword to let users opt-in/out, and so that pandas can propogate its ``use_nullable_dtypes`` keyword down to fastparquet. 

The default should probably be true in order to maintain backwards compatibility.

**Comment by martindurant** [comment]
Fastparquet would like to make use of this pandas feature, which greatly improves efficiency, simplifies the code, and makes true roundtrips of data possible. The nullable types map exactly to parquet types.

I would argue, that pandas should make, for example `"uint8" == "UInt8"` be True. That would solve your tests. I don't think fastparquet wants to go back to casting to float, even if pandas still does that in the DataFrame constructor.

**Comment by martindurant** [comment]
(PS: nullable types have been around since pandas 1.0.0. I know the doc says experimental, but not really!)

**Comment by lithomas1** [comment]
Regarding the uint8 case, that is probably a bug in fastparquet.
See this code snippet(tested on master, but probably reproduces on 1.3 also).
```
>>> import pandas as pd
>>> a = pd.DataFrame({"a":[1,2,3]}, dtype="uint8")
>>> a
   a
0  1
1  2
2  3
>>> a.dtypes
a    uint8
dtype: object
>>> a.to_parquet("test.parquet")
>>> pd.read_parquet("test.parquet")
   a
0  1
1  2
2  3
>>> pd.read_parquet("test.parquet").dtypes
a    UInt8
dtype: object
```
The dtype should be uint8 since there are no NA values.

Regarding making ``"uint8"=="Uint8"`` be true, the only case that will happen is if there are no nulls. If there are no nulls, we should always want ``"uint8"`` since the nullable dtypes pretty much double memory usage because of the mask generated. So, I think pandas should be strict in testing that uint8 dtypes are used when no nulls are present b/c of the performance impact.

**Comment by martindurant** [comment]
I agree that your example should return uint8, I'll fix it.

> If there are no nulls, we should always want "uint8"

You do not know this until you have loaded the data, unfortunately.

**Comment by martindurant** [comment]
#643

**Comment by lithomas1** [comment]
> > If there are no nulls, we should always want "uint8"
> 
> You do not know this until you have loaded the data, unfortunately.
I meant that uint8 shouldn't be equal to Uint8 only in testing. It did catch the bug fixed in #643, so I would argue its pretty useful. From a user perspective, though, I would probably want them to behave pretty much the same.

AFAIK, the nullable dtypes are experimental since we don't support it a lot of methods, and because there are some nasty bugs, such as with the nullable integer dtype mangling large integers. So, because of its experimental nature, pandas makes using them opt-in in its I/O methods. (I think it'll probably get promoted out of experimental around 2.0, but I'm not sure as I'm not on the core team.)

We do appreciate fastparquet trying out our new types, but unfortunately, without a way to opt-in/out, I don't think that we can support the new fastparquet(I think pandas users expect the float casting, and we have to maintain backwards compatibility). Perhaps, a good compromise here would be to have the option turn the nullable dtypes on by default, with an option to toggle off? Users would still see the perf benefit, and compat with ``pandas.read_parquet`` would be restored. However, you would still have to maintain the casting to float code.

Please LMK what you think.

**Comment by jorisvandenbossche** [comment]
It's of course up to you, but I would personally recommend to reconsider making them opt-in (so switching the default of the keyword).

> (PS: nullable types have been around since pandas 1.0.0. I know the doc says experimental - for integers anyway - but not really!)

How much I wished that they were already more stable (as one of the main driving forces behind them), there are really still a lot of operations in pandas that are not or poorly supported with nullable dtypes, as @lithomas1 already mentioned above as well. One other simple example: setting such a column as the index is not yet supported and will give you an object dtype index ..

I also do appreciate that people (and packages) try it out (so we get bug reports about it etc), but I think at this point that's best decided by the end user. After some more time, `pandas_nulls=True` could again become the default.
